### PR TITLE
Improve compress for cuda-aware mpi

### PR DESCRIPTION
**Pull request by Rombur**
This PR does two things:
 - update the api of a few functions in `cuda_kernel` for consistency (these functions are a couple of months old so we are free to change the API)
 - use the same logic found in #7303 (where it was used to speed up `update_ghost`) to speed up `compress`

cc: @dsambit

### CODE REVIEW

**Comment by tamiko** [comment]
(It is probably a moot point to let the CI run, but let's check the non-cuda scenario for successful compilation anyway.)

File: include/deal.II/base/partitioner.templates.h [hunk 1]
```diff
@@ -633,70 +607,90 @@ namespace Utilities
 #    else
           if (vector_operation == dealii::VectorOperation::add)
             {
-              for (const auto &import_range : import_indices_data)
+              const unsigned int import_indices_plain_dev_size =
+                import_indices_plain_dev.size();
+              for (unsigned int i = 0; i < import_indices_plain_dev_size; ++i)
```
Comment (masterleinad) [inline include/deal.II/base/partitioner.templates.h]: It seems that you can still use a range-based for loop here and I expect `clang-tidy` to complain if it sees this loop.

File: include/deal.II/base/partitioner.templates.h [hunk 2]
```diff
@@ -633,70 +607,90 @@ namespace Utilities
...
-                  <<<n_blocks, dealii::CUDAWrappers::block_size>>>(
-                    locally_owned_array.data() + import_range.first,
-                    read_position,
-                    chunk_size);
-                read_position += chunk_size;
-              }
+            {
+              const unsigned int import_indices_plain_dev_size =
+                import_indices_plain_dev.size();
+              for (unsigned int i = 0; i < import_indices_plain_dev_size; ++i)
```
Comment (masterleinad) [inline include/deal.II/base/partitioner.templates.h]: see above

File: include/deal.II/base/partitioner.templates.h [hunk 3]
```diff
@@ -633,70 +607,90 @@ namespace Utilities
...
-                             read_position,
-                             chunk_size * sizeof(Number),
-                             cudaMemcpyDeviceToDevice);
-                AssertCuda(cuda_error_code);
-                read_position += chunk_size;
-              }
+            {
+              const unsigned int import_indices_plain_dev_size =
+                import_indices_plain_dev.size();
+              for (unsigned int i = 0; i < import_indices_plain_dev_size; ++i)
```
Comment (masterleinad) [inline include/deal.II/base/partitioner.templates.h]: see above

File: include/deal.II/base/partitioner.templates.h [hunk 4]
```diff
@@ -633,70 +607,90 @@ namespace Utilities
...
+                      read_position,
+                      chunk_size);
+                  read_position += chunk_size;
+                }
+            }
+          else
+            {
+              const unsigned int import_indices_plain_dev_size =
+                import_indices_plain_dev.size();
+              for (unsigned int i = 0; i < import_indices_plain_dev_size; ++i)
```
Comment (masterleinad) [inline include/deal.II/base/partitioner.templates.h]: see above

File: include/deal.II/lac/cuda_kernels.templates.h [hunk 5]
```diff
@@ -573,9 +591,9 @@ namespace LinearAlgebra
 
       template <typename Number>
       __global__ void
-      add_permutated(Number *         val,
+      add_permutated(const size_type *indices,
```
Comment (masterleinad) [inline include/deal.II/lac/cuda_kernels.templates.h]: Why doesn't have this function an `IndexType` template parameter? Would it make sense for conformity?

Comment (Rombur) [reply include/deal.II/lac/cuda_kernels.templates.h]: Because it doesn't need to be templated. I only template the functions that have to be templated

Comment (masterleinad) [reply include/deal.II/lac/cuda_kernels.templates.h]: OK

**Comment by masterleinad** [comment]
I can push the range-based loop changes here or create a PR afterward if you prefer that.

**Comment by Rombur** [comment]
> I can push the range-based loop changes here or create a PR afterward if you prefer that.

That's fine, I'll do it.

**Comment by Rombur** [comment]
@masterleinad done

**Comment by masterleinad** [comment]
Passes all `cuda` tests for me locally.

File: include/deal.II/base/partitioner.h [hunk 6]
```diff
@@ -571,6 +571,13 @@ namespace Utilities
                      << " elements for this partitioner.");
 
     private:
+      /**
+       * Initialize import_indices_plain_dev from import_indices_data. This
+       * function is only used when CUDA-aware MPI.
```
Comment (bangerth) [inline include/deal.II/base/partitioner.h]:
```suggestion
       * function is only used when using CUDA-aware MPI.
```

**Pull request merged**